File: FOXYLQVH34LJ9/bin/python/fixidea/main.py
#!/usr/bin/env pipenv run python

import sys
import os
import re
import xml.etree.ElementTree as ET

sourceFolders = [
    {
        'url': 'file://$MODULE_DIR$/src',
        'isTestSource': 'false',
        'packagePrefix': 'App\\',
    },
    {
        'url': 'file://$MODULE_DIR$/tests',
        'isTestSource': 'true',
        'packagePrefix': 'App\\Tests\\',
    },
    {
        'url': 'file://$MODULE_DIR$/vendor/gdbots',
        'isTestSource': 'false',
    },
    {
        'url': 'file://$MODULE_DIR$/vendor/triniti',
        'isTestSource': 'false',
    },
]

excludeFolders = [
    'file://$MODULE_DIR$/vendor/gdbots/schemas',
    'file://$MODULE_DIR$/vendor/triniti/schemas',
]

def parseConfig(configFile):
    repoDir = os.path.dirname(os.path.dirname(configFile))
    repo = os.path.basename(repoDir)
    if not (re.match(r"(tmz|toofab)-(web|api|feeds|share|amp)", repo)):
        print(f'Invalid repo: {repo}')
        return

    vendor = repo.split('-')[0]
    sourceFolders.append({
        'url': f'file://$MODULE_DIR$/vendor/{vendor}',
        'isTestSource': 'false',
    })

    excludeFolders.append(f'file://$MODULE_DIR$/vendor/{vendor}/schemas')
        
    tree = ET.parse(configFile)
    root = tree.getroot()
    for component in root.findall('component'):
        for content in component.findall('content'):
            for sourceFolder in content.findall('sourceFolder'):
                content.remove(sourceFolder)

            for excludeFolder in content.findall('excludeFolder'):
                content.remove(excludeFolder)

            for sourceFolder in sourceFolders:
                el = ET.SubElement(content, 'sourceFolder')
                el.set('url', sourceFolder['url'])
                el.set('isTestSource', sourceFolder['isTestSource'])
                if ('packagePrefix' in sourceFolder):
                    el.set('packagePrefix', sourceFolder['packagePrefix'])

            el = ET.SubElement(content, 'excludeFolder')
            el.set('url', 'file://$MODULE_DIR$/var')

            if (os.path.isdir(f'{repoDir}/public')):
                el = ET.SubElement(content, 'excludeFolder')
                el.set('url', 'file://$MODULE_DIR$/public/client')

            if not (os.path.isdir(f'{repoDir}/vendor')):
                continue

            for excludeFolder in excludeFolders:
                el = ET.SubElement(content, 'excludeFolder')
                el.set('url', excludeFolder)

            for dependency in os.listdir(f'{repoDir}/vendor'):
                if (dependency == vendor):
                    continue
                if not (re.match(r"(gdbots|triniti)", dependency)):
                    el = ET.SubElement(content, 'excludeFolder')
                    el.set('url', f'file://$MODULE_DIR$/vendor/{dependency}')

    ET.indent(tree, '  ')
    tree.write(configFile, encoding='utf-8', xml_declaration=True)

def main():
    if not os.path.isfile(sys.argv[1]):
        print('file does not exist')
        return
    parseConfig(sys.argv[1])

if __name__ == "__main__":
    main()

File: FOXYLQVH34LJ9/bin/python/fixidea/main.mjs
import fs from 'fs';
import path from 'path';
import { DOMParser, XMLSerializer } from 'xmldom';

const sourceFolders = [
  {
    url: 'file://$MODULE_DIR$/src',
    isTestSource: 'false',
    packagePrefix: 'App\\',
  },
  {
    url: 'file://$MODULE_DIR$/tests',
    isTestSource: 'true',
    packagePrefix: 'App\\Tests\\',
  },
  {
    url: 'file://$MODULE_DIR$/vendor/gdbots',
    isTestSource: 'false',
  },
  {
    url: 'file://$MODULE_DIR$/vendor/triniti',
    isTestSource: 'false',
  },
];

const excludeFolders = [
  'file://$MODULE_DIR$/vendor/gdbots/schemas',
  'file://$MODULE_DIR$/vendor/triniti/schemas',
];

function parseConfig(configFile) {
  const repoDir = path.dirname(path.dirname(configFile));
  const repo = path.basename(repoDir);
  if (!/(tmz|toofab)-(web|api|feeds|share|amp)/.test(repo)) {
    console.log(`Invalid repo: ${repo}`);
    return;
  }

  const vendor = repo.split('-')[0];
  sourceFolders.push({
    url: `file://$MODULE_DIR$/vendor/${vendor}`,
    isTestSource: 'false',
  });

  excludeFolders.push(`file://$MODULE_DIR$/vendor/${vendor}/schemas`);

  fs.readFile(configFile, 'utf8', (err, data) => {
    if (err) {
      console.error(err);
      return;
    }

    const doc = new DOMParser().parseFromString(data, 'application/xml');
    const components = doc.getElementsByTagName('component');

    for (let i = 0; i < components.length; i++) {
      const component = components[i];
      const contents = component.getElementsByTagName('content');

      for (let j = 0; j < contents.length; j++) {
        const content = contents[j];

        // Remove existing sourceFolder and excludeFolder elements
        const sourceFolders = content.getElementsByTagName('sourceFolder');
        while (sourceFolders.length > 0) {
          content.removeChild(sourceFolders[0]);
        }

        const excludeFolders = content.getElementsByTagName('excludeFolder');
        while (excludeFolders.length > 0) {
          content.removeChild(excludeFolders[0]);
        }

        // Add new sourceFolder elements
        sourceFolders.forEach(sourceFolder => {
          const el = doc.createElement('sourceFolder');
          el.setAttribute('url', sourceFolder.url);
          el.setAttribute('isTestSource', sourceFolder.isTestSource);
          if (sourceFolder.packagePrefix) {
            el.setAttribute('packagePrefix', sourceFolder.packagePrefix);
          }
          content.appendChild(el);
        });

        // Add new excludeFolder elements
        const varExclude = doc.createElement('excludeFolder');
        varExclude.setAttribute('url', 'file://$MODULE_DIR$/var');
        content.appendChild(varExclude);

        if (fs.existsSync(`${repoDir}/public`)) {
          const publicExclude = doc.createElement('excludeFolder');
          publicExclude.setAttribute('url', 'file://$MODULE_DIR$/public/client');
          content.appendChild(publicExclude);
        }

        if (!fs.existsSync(`${repoDir}/vendor`)) {
          continue;
        }

        excludeFolders.forEach(excludeFolder => {
          const el = doc.createElement('excludeFolder');
          el.setAttribute('url', excludeFolder);
          content.appendChild(el);
        });

        fs.readdirSync(`${repoDir}/vendor`).forEach(dependency => {
          if (dependency === vendor) {
            return;
          }
          if (!/(gdbots|triniti)/.test(dependency)) {
            const el = doc.createElement('excludeFolder');
            el.setAttribute('url', `file://$MODULE_DIR$/vendor/${dependency}`);
            content.appendChild(el);
          }
        });
      }
    }

    const xml = new XMLSerializer().serializeToString(doc);
    fs.writeFile(configFile, xml, 'utf8', err => {
      if (err) {
        console.error(err);
      }
    });
  });
}

function main() {
  const configFile = process.argv[2];
  if (!fs.existsSync(configFile)) {
    console.log('file does not exist');
    return;
  }
  parseConfig(configFile);
}

main();